src/forecast.py

```diff
@@ -213,8 +199,112 @@ def forecast_page():
+                full_time_df = pd.DataFrame(full_time_range, columns=['start_time'])
+                forecast = full_time_df.merge(forecast, on='start_time', how='left')
+
+                fig.add_trace(go.Scatter(
+                    x=forecast["start_time"],
+                    y=forecast["quantity"],
+                    mode='lines',
+                    name=f"Elexon {process_types[i]}",
+                    line=dict(color='#318CE7', dash=line_style),
+                    connectgaps=False
+                ))
+
+            st.plotly_chart(fig, theme="streamlit")
+        else:
+            st.error("Invalid date range. Start date must be before end date.")
+
+# Function to fetch and process data
+def fetch_forecast_data(api_func, start_date, end_date, process_type):
+    try:
+        response = api_func(
+            _from=start_date.isoformat(),
+            to=end_date.isoformat(),
+            process_type=process_type,
+            format="json",
+        )
+        if not response.data:
+            return pd.DataFrame()
+
+        df = pd.DataFrame([item.to_dict() for item in response.data])
+        solar_df = df[df["business_type"] == "Solar generation"]
+        solar_df["start_time"] = pd.to_datetime(solar_df["start_time"])
+        solar_df = solar_df.set_index("start_time")
+
+        # Only resample if there's data
+        if not solar_df.empty:
+            solar_df = solar_df.resample("30T")["quantity"].sum().reset_index()
+
+        return solar_df
+    except Exception as e:
+        st.error(f"Error fetching data for process type '{process_type}': {e}")
+        return pd.DataFrame()
+
+def determine_start_and_end_datetimes(start_datetimes, end_datetimes):
+    """
+    Determines the start and end datetime in UTC.
+    Parameters:
+    - start_datetimes: list of datetime or date objects
+    - end_datetimes: list of datetime or date objects
+    Returns:
+    - start_datetime_utc: datetime object in UTC
+    - end_datetime_utc: datetime object in UTC
+    """
+    now = datetime.utcnow()
+
+    # Determine start_datetime_utc
+    if start_datetimes:
+        start_datetime_utc = start_datetimes[0]
+    else:
+        start_datetime_utc = now
+
+    # Ensure start_datetime_utc is a datetime object
+    if isinstance(start_datetime_utc, date) and not isinstance(start_datetime_utc, datetime):
+        start_datetime_utc = datetime.combine(start_datetime_utc, datetime.min.time())
+
+    # Determine end_datetime_utc
+    if end_datetimes and end_datetimes[-1]:
+        end_datetime_utc = end_datetimes[-1]
+    else:
+        end_datetime_utc = start_datetime_utc + timedelta(days=7)
+
+    # Ensure end_datetime_utc is a datetime object
+    if isinstance(end_datetime_utc, date) and not isinstance(end_datetime_utc, datetime):
+        end_datetime_utc = datetime.combine(end_datetime_utc, datetime.min.time())
+
+    # Check if start is before end
+    if start_datetime_utc >= end_datetime_utc:
```

Comment:
do we need this? could we do an assert instead? then we dont need the `if` statement above